test/test_type_promotion.py

```diff
@@ -138,39 +139,50 @@ def test_mixed_type_backward(self):
-    # where that should hold.
+        for func in [torch.add, torch.sub, torch.rsub, torch.mul, torch.div]:
+            for dtype1, dtype2 in itertools.product(f_dtypes, f_dtypes + i_dtypes):
+                x = torch.ones(10, requires_grad=True, dtype=dtype1, device=self.device)
+                y = torch.ones(10, dtype=dtype2, device=self.device)
+                func(x, y).sum().backward()
+
+    # verifies that torch.<op>(first, second) is the same as 
+    # torch.<op>(first.to(common_dtype), second.to(common_dtype)) in cases where that should hold.
     def test_many_promotions(self):
-        from_to = {
-            torch.float16: torch.float32,
-            torch.half: torch.float16,
-            torch.int: torch.long,
-            torch.uint8: torch.long,
-            torch.uint8: torch.float,
-            torch.int: torch.float,
-            torch.int: torch.double,
-            torch.int16: torch.long,
-            torch.float16: torch.double,
-            torch.bool: torch.long,
-            torch.bool: torch.float
-        }
-
-        for k, v in from_to.items():
-            a = torch.rand([3, 3], device=self.device).to(k)  # no _th_uniform for half on cpu.
-            b = torch.rand([3, 3], device=self.device).to(v)
-            c = a.add(b)
-            d = a.to(v).add(b)
-            self.assertEqual(c.dtype, d.dtype, message='from {} to {}'.format(k, v))
-            self.assertEqual(c, d, message='from {} to {}'.format(k, v))
+        # Can also include half on CPU in cases where it will be promoted to a
+        # supported dtype
+        dtypes1 = torch.testing.get_all_math_dtypes('cuda')
+        dtypes2 = torch.testing.get_all_math_dtypes(self.device)
+        ops = [torch.add, torch.sub, torch.mul, torch.div, torch.rsub]
+        for dt1, dt2 in itertools.product(dtypes1, dtypes2):
+            for op, non_contiguous in itertools.product(ops, [True, False]):
+                common_dtype = torch.promote_types(dt1, dt2)
+                if common_dtype == torch.half:
+                    continue
+                if op == torch.sub and common_dtype != torch.bool:
+                    # Subtraction, the `-` operator, with a bool tensor is not supported.
+                    continue
+                first = torch.rand([20, 20, 20], device=self.device) * 100
+                second = torch.rand([20, 20, 20], device=self.device) * 100
+                # ensure some zeros, otherwise boolean version is all true
+                first[first < -90] = 0
+                first = first.to(dt1)
+                second = second.to(dt2)
+                # ensure no div-by-zero
+                second[second == 0] = 5
+                # test ops with non-contiguous tensors
+                if non_contiguous:
+                    first = first.transpose(0, 2)
+                    second = second.transpose(2, 1)
+                    # ensure we can catch non-contiguous issues that might only arise when tensors 
```

Comment:
nit: this can be tossed into the error message for self.assertNotEqual instead of being a comment.